Code/src/main/antlr4/nl/utwente/group10/haskell/typeparser/Type.g4

```diff
@@ -4,14 +4,20 @@ CT : [A-Z][a-z]+ ;
 VT : [a-z]+ ;
 WS : [ \t\r\n]+ -> skip ;
 
-type : functionType | compoundType ; // function type
-functionType : compoundType '->' type ;
+type : typeClasses ? innerType ;
+innerType : functionType | compoundType ; // function type
+functionType : compoundType '->' innerType ;
 compoundType : constantType | variableType | tupleType | listType | parenType ;
 
-tupleType : '(' type (',' type)+ ')' ; // tuple type, k>=2
-listType : '[' type ']' ;              // list type
-parenType : '(' type ')' ;             // type with parentheses
+tupleType : '(' innerType (',' innerType)+ ')' ; // tuple type, k>=2
+listType : '[' innerType ']' ;              // list type
+parenType : '(' innerType ')' ;             // type with parentheses
 
-constantType : typeConstructor (WS* type)* ;
+constantType : typeConstructor (WS* innerType)* ;
 typeConstructor : CT ;
 variableType : VT ;
+
+typeClasses : '(' typeWithClass (',' typeWithClass)* ')' '=>' ;
```

Comment:
I just found out that if there is only one typeclass constraint, you can also write

    f :: Num a => a
    f = 10

That is, without the parentheses. The typeClasses rule would then become something like

    typeClasses : '(' tWC (',' tWC)+ ')' '=>' | tWC '=>' ;

I think.